File: java/src/main/java/nl/javadude/codejam/gcj08/qualify/SavingTheUniverse.java
package nl.javadude.codejam.gcj08.qualify;

import java.io.BufferedReader;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import nl.javadude.codejam.AbstractMultiProblemSolver;

public class SavingTheUniverse extends AbstractMultiProblemSolver<SavingTheUniverseProblem> {

	@Override
	protected SavingTheUniverseProblem readProblem(BufferedReader lines, int problemNumber) throws Exception {
		SavingTheUniverseProblem problem = new SavingTheUniverseProblem();
		problem.problemNumber = problemNumber;
		problem.nrEngines = readLineAsInt(lines);
		readLinesToList(lines, problem.nrEngines, problem.engines);
		problem.nrSearches = readLineAsInt(lines);
		readLinesToList(lines, problem.nrSearches, problem.searches);
		return problem;
	}

	@Override
	protected List<String> solveProblem(SavingTheUniverseProblem problem) {
		int switches = 0;
		Set<String> enginesEncountered = new HashSet<String>();
		for (String search : problem.searches) {
			enginesEncountered.add(search);
			if (enginesEncountered.size() == problem.nrEngines) {
				switches++;
				enginesEncountered.clear();
				enginesEncountered.add(search);
			}
		}
		
		return Collections.singletonList("Case #" + problem.problemNumber + ": " + 	switches);
	}

}

File: src/main/scala/nl/javadude/codejam/gcj08/qualify/SavingTheUniverse.scala
package nl.javadude.scala.gcj08.qualify
import nl.javadude.scala.utils.GoogleCodeHelper._
import nl.javadude.scala.CodeJam

/**
 * VERIFIED against Google Code Jam
 * @author Jeroen van Erp - jeroen at hierynomus.com
 */
object SavingTheUniverse extends CodeJam {

  override def solveProblem(reader: Iterator[String]) : String = {
	  val nrEngines = reader.nextInt
	  reader skipLines nrEngines
	  var enginesEncountered = Set[String]()
	  var switches = 0
	  for (val s <- 1 to reader.nextInt) {
		  val read = reader.next
		  enginesEncountered += read
		  if (enginesEncountered.size == nrEngines) {
			  enginesEncountered = Set(read)
			  switches += 1
		  }
	  }
	  switches.toString
  }
}
